src/lib/collections/index.ts

```diff
@@ -0,0 +1,86 @@
+import { getCollection as getAstroCollection, getEntry as getAstroCollectionEntry } from 'astro:content';
+import { PUBLIC_IS_PRODUCTION, HEAD_START_PREVIEW } from 'astro:env/server';
+import pagesCollection from './pages';
+import type { SiteLocale } from '@lib/i18n/types';
+import { combine, split } from './lib/slug';
+
+export const collectionMap = {
+  ...pagesCollection,
+  // Add more collectionMap here
+} as const;
+
+const useLoaderDirectly = !PUBLIC_IS_PRODUCTION || HEAD_START_PREVIEW;
+
+type CollectionKey = keyof typeof collectionMap;
+type CollectionEntry<Key extends CollectionKey> = NormalizedEntry<
+  Awaited<ReturnType<typeof collectionMap[Key]['loader']>>[number]
+>;
+
+/**
+ * Fetches entries from a collection.
+ * 
+ * In development and preview environments, data is loaded directly using the collection loader.
+ * In production environments, data is retrieved using Astro's getCollection function.
+ * 
+ * @param collection - The key of the collection to fetch entries from
+ * @param filter - Optional SiteLocale or function to filter the collection entries
+ * @returns A promise that resolves to an array of normalized collection entries
+ */
+export async function getCollection<Key extends CollectionKey>(
+  collection: Key,
+  filter?: SiteLocale | ((entry: CollectionEntry<Key>) => boolean),
+): Promise<CollectionEntry<Key>[]> {
+  const filterFunction = (typeof filter === 'string') 
+    ? ({ id }: CollectionEntry<Key>) => {
+      const { locale } = split(id);
+      return !locale || locale === filter;
+    }
+    : filter;
+  
+  return (useLoaderDirectly) 
+    ? collectionMap[collection].loader()
+      .then(entries => normalizeEntries(entries, collection))
```

Comment:
```suggestion
      .then(entries => normalizeEntries<typeof entries[number]>(entries, collection))
```
Without it, entries might become a union of collection entries.